siac/main.go

```diff
@@ -1,101 +1,117 @@
 package main
 
 import (
+	"encoding/json"
+	"net/http"
 	"os"
+	"reflect"
 
 	"github.com/spf13/cobra"
 )
 
-func versioncmd(*cobra.Command, []string) {
-	println("Sia Client v0.1.0")
+const (
+	VERSION  = "0.2.0"
+	hostname = "http://localhost:9980"
+)
+
+// getAPI makes an API call and decodes the response.
+func getAPI(call string, obj interface{}) (err error) {
+	resp, err := http.Get(hostname + call)
+	if err != nil {
+		return
+	}
+	err = json.NewDecoder(resp.Body).Decode(obj)
+	if err != nil {
+		return
+	}
+	resp.Body.Close()
+	return
+}
+
+// callAPI makes an API call and discards the response.
+func callAPI(call string) (err error) {
+	resp, err := http.Get(hostname + call)
+	if err != nil {
+		return
+	}
+	resp.Body.Close()
+	return
+}
+
+// wrap wraps a generic command with a check that the command has been
+// passed the correct number of arguments. The command must take only strings
+// as arguments.
+func wrap(fn interface{}) func(*cobra.Command, []string) {
+	fnVal, fnType := reflect.ValueOf(fn), reflect.TypeOf(fn)
+	if fnType.Kind() != reflect.Func {
+		panic("wrapped function has wrong type signature")
+	}
+	for i := 0; i < fnType.NumIn(); i++ {
+		if fnType.In(i).Kind() != reflect.String {
+			panic("wrapped function has wrong type signature")
+		}
+	}
+
+	return func(cmd *cobra.Command, args []string) {
+		if len(args) != fnType.NumIn() {
+			cmd.Usage()
+			return
+		}
+		argVals := make([]reflect.Value, fnType.NumIn())
+		for i := range args {
+			argVals[i] = reflect.ValueOf(args[i])
+		}
+		fnVal.Call(argVals)
+	}
 }
 
 func main() {
 	root := &cobra.Command{
 		Use:   os.Args[0],
-		Short: "Sia Client v0.1.0",
-		Long:  "Sia Client v0.1.0",
-		Run:   versioncmd,
+		Short: "Sia Client v" + VERSION,
+		Long:  "Sia Client v" + VERSION,
+		Run:   wrap(func() { println("Sia Client v" + VERSION) }),
```

Comment:
I'm not sure that I view this abstraction as a good thing. It's pretty intense in terms of complexity.